Explicitly disallow multiple Qt bindings packages in a frozen application

rokm opened the pull request:
Explicitly disallow multiple Qt bindings packages in a frozen application. 

Due to the way that collected Qt environment is set up (the paths set up by embedded `qt.conf` resource that is injected either via bindings packages themselves, or by our run-time hooks, as well as environment variables set by our run-time hooks), only one Qt bindings package can be properly configured (i.e., the paths point to Qt directories belonging to that package). Other binding packages usually end up broken, and fail to load either due to incompatible shared libraries or missing (or rather, unloadable) Qt platform plugins.

Therefore, we now raise an error if we detect attempts to run hooks for more than one Qt bindings package. This is done both at build time (in hooks for top-level bindings packages) as well as run-time (in run-time hooks for Qt bindings packages).

rokm committed:
rthooks: qt: allow only one Qt bindings run-time hook to be ran

Have the run-time hooks for Qt bindings check whether any other
run-time hook for Qt bindings has been run before, and raise an
error if it has.

This is because due to all effort that goes into making the
collected Qt bindings package relocatable via the embedded
`qt.conf` resource (either injected by packages themselves, or
by our run-time hooks) and environment variables set by the
run-time hooks, there can only be one properly configured Qt
bindings package in the frozen application.

So better to crash early with an error message that describes the
problem than crashing later with a message about incompatible
shared library or missing/non-loadable platform plugins.

rokm committed:
hooks: qt: allow only one Qt bindings package to be collected

Have the hook for top-level Qt bindings packages check whether a
top-level hook for different Qt binding package has already been
run. If it has, abort the build process with error that tells user
to either ensure there is only one Qt bindings package installed in
the build environment, or use the module exclusion mechanism to
exclude extraneous package(s).

rokm commented:
Hmm, I guess at least we know that the check works as intended...

bwoodsend commented:
So those are all libraries which (indirectly) suck in multiple Qt variants? That's unreasuring.

rokm commented:
Yeah, I'll take a closer look at import chains - I wouldn't be surprised if it is `matplotlib` that's pulling them in, though.

rokm committed:
hookutils: qt: implement helper for excluding extraneous Qt bindings

Implement new helper, called `exclude_extraneous_qt_bindings`,
for generating lists of excludes for suppressing collection
of extraneous Qt bindings. It is intended for use in hooks for
modules that conditionally import all Qt bindings packages, and
thus end up pulling all available bindings into the frozen
application.

The helper attempts to determine Qt bindings to use by:

1. checking if a hook for top-level bindings package has been run;
   if it has, select those bindings. This covers the cases when
   entry-point script explicitly imports bindings before importing
   problematic modules.

2. check the `QT_API` environment variable; if it contains a valid
   Qt bindings name, select those bindings (without checking if
   they are actually available).

3. pick one of available bindings, and display a warning if
   multiple bindings are available.

After determining the bindings to use, return list of all other
bindings (their top-level package names), which the hook should
assing to its `excludedimports` hook variable.

rokm committed:
hooks: matplotlib: prevent collection of multiple Qt bindings

Add a hook for `matplotlib.backends.qt_compat` module that uses
the new `exclude_extraneous_qt_bindings` helper function to
select Qt bindings to use and exclude extraneous Qt bindings.

Add hooks for `matplotlib.backends.backend_qtagg` and
`matplotlib.backends.backend_qtcairo` to suppress their conditional
imports of `PyQt5`, to prevent the latter from being pulled in
along with the bindings selected by the `qt_compat` hook.

rokm committed:
hooks: qtpy: prevent collection of multiple Qt bindings

Add a hook for `qtpy` package that uses the new
`exclude_extraneous_qt_bindings` helper function to select Qt
bindings to use and exclude extraneous Qt bindings.

rokm committed:
hooks: pandas: exclude conditional import of PyQt5

Add a hook for `pandas.io.clipboard` package to exclude the
conditional import of `PyQt5` that the package makes. This import
is done as a fall-back if `qtpy` is unavailable, however, the
contemporary versions of `pandas` require both `qtpy` and `PyQt5`
for their `clipboard` and `all` extras.

So the `PyQt5` import made in `pandas.io.clipboard` should be
suppressed, to avoid collecting it along with whatver Qt bindings
is selected by our hook for `qtpy`.

rokm commented:
Looks like https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/708 will be pre-requisite for the CI passing here. 

We also have a `pyqtgraph` hook in contrib repository, and that will likely also have to be updated to use the Qt backend selection helper (when PyInstaller >= 6.5)...

bwoodsend commented:
I'm not sure we can get away with not calling this a breaking change. I imagine that most Qt using libraries do similar Qt variant abstractions.

rokm commented:
> I'm not sure we can get away with not calling this a breaking change. I imagine that most Qt using libraries do similar Qt variant abstractions.

True. I'll add a breaking change news entry with a note at the end that any packages leading to problems with this should be reported, so that we can hook them up.

rokm committed:
analysis: try to pre-load hook as soon as module is encountered

Try to pre-load hooks as soon as their corresponding module is
encountered during the analysis, in an attempt to ensure that
hooks are loaded in roughly the same order as imports are
encountered during the analysis.

This behavior is implicitly required by the new helper
`PyInstaller.utils.hooks.qt.exclude_extraneous_qt_bindings`
for automatic selection of Qt bindings, as it attempts to
infer the preferred Qt bindings based on hooks that have
already been run.

Our hooks are usually loaded in the post-graph analysis phase,
and the order there does not correspond to order of imports
in the code. However, even prior to this commit, some hooks
have been loaded during the analysis stage as well. Specifically,
when a module is imported from another module, our `_safe_import_hook`
implementation checks for exclusion rules in hooks for that
referring (source) module and its parent package(s). This loads
the hooks in question, but because they are cached, they are also
re-used in post-graph processing without being loaded again.

So we augment the hook loading in `_safe_import_hook` by adding
a pre-load step after calling the parent implementation and before
returning the result (i.e., the list of modules that have been
imported).

rokm commented:
Another problem came up while I was testing modified `pyqtgraph` hook in contributed hooks repository. 

Turns out the hooks are not loaded in the order that imports are encountered during the analysis. This is because hooks are typically loaded during post-graph analysis, where we loop over set of hooks and look for associated modules in modulegraph. However, we sometimes do load a hook during analysis itself (and effectively pre-load it for post-graph analysis due to hook caching) - when a module is imported from another module, we check for exclusion rules of that referring (source) module and its parent package(s), which means that we need to load their hooks.

So as a sort of work-around, I've added an explicit attempt to (pre)load hooks for target modules at the end of `_safe_import_hook`, before we return the list. Due to hook caching, this should not have any performance implications. It's not especially pretty, but it does ensure that if user imports particular Qt bindings before `pyqtgraph` in their entry-point script, the hook for bindings is called before the hook for `pyqtgraph`, and consequently the `exclude_extraneous_qt_bindings` in the latter can correctly identify the bindings that should be used.
